README
===========================
dl: download ticket service
===========================

dl is a file exchange service that allows you to upload any file to a web
server and generate a unique ticket for others to download. The ticket is
automatically expired according to the specified rules, so that you don't need
to keep track or cleanup afterward. dl also allows you to grant an anonymous,
one-time upload for others to send *you* a file, without the requirement of
account management.

dl is usually installed as a "email attachments replacement" due to its
simplicity (though can be used in other ways).

.. contents::


Requirements
============

* PHP 5.0 or higher.
* PHP sqlite module (or another PDO database module).
* Web server access for installation and setup.

Optionals:

* PHP APC module, for the upload progress-bar.


Installation
============

The installation is divided into three steps:

1) Installation of the files
2) Database creation/setup or database upgrade
3) Initial user creation

Due to the number of possible configurations, the installation must be carried
on manually.


DL installation
---------------

* Copy the ``htdocs`` directory contained in the archive to a directory of
  choice under your web server.

* Copy ``include/config.php.dist`` to either ``include/config.php`` or
  ``/etc/dl.php`` and customize as needed. This usually means changing the URL
  and E-Mail of the service.

* Ensure the ``include`` and ``style/include`` directories are *not* directly
  accessible externally. Type ``http://dl.example.com/include/config.php`` and
  verify that you *correctly* get an "Access Denied" error.

  If you use Apache, the provided ``.htaccess`` files should be already
  sufficient; consult your web server documentation otherwise.

* Create a spool directory that will be used by the service to store the files,
  user and ticket information. This directory *should* be outside of the
  document root of your web server. Fix the permissions so that PHP has
  read/write access to it.

  In the provided ``include/config.php`` this is configured as
  ``/var/spool/dl``. If you web server runs as "nobody:nogroup", issue::

    mkdir -p -m770 /var/spool/dl
    chgrp nogroup /var/spool/dl

  to create correctly this directory.

* Create a database (as described in `Database setup`_).

* Create at least one user (as described in `Internal authentication`_).


Database setup
--------------

DL needs a database to store the ticket and user information. By default, DL
will use an embedded "sqlite" database stored within the spool directory, but
some manual setup is still required.

To create the database, you need the ``sqlite`` or ``sqlite3`` command,
and execute the provided SQL batch for sqlite *with the same user as your web
server* (by using ``su`` or ``sudo`` if necessary)::

  cd /var/spool/dl/
  sqlite data.sdb < /your-installation-directory/include/scripts/sqlite.sql
  chmod 660 data.sdb

If you want to change the database path, or use a real database server, you
need to properly configure the ''$dsn'' parameter in ``include/config.php``
according to your setup.


Upgrading from DL <= 0.3
~~~~~~~~~~~~~~~~~~~~~~~~

Upgrading from DL 0.3 or earlier is possible without loosing any data by using
the bundled ``include/scripts/dbupgrade.php`` script. Proceed as follows:

* Backup your current ``config.php`` file and spool directory.

* Overwrite the DL installation directory with the new copy.

* Customize the new ``config.php`` file.

* Create a new database as described above in `Database setup`_.

* If you had to change the ''$dbHandler'' parameter in DL 0.3,
  copy the same setting into the new configuration file.

* Run ``dbupgrade.php`` as your web server user::

    cd /your-installation-directory/include/scripts
    php dbupgrade.php

* Test your new setup.

* You must now remove ''$dbHandler'' from the new configuration file and
  ``data.db``, ``user.db`` from the spool directory.


Upgrading from DL 0.3 and any later version
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

* Backup your current ``config.php`` file and spool directory.

* Overwrite the DL installation directory with the new copy.

* Either copy over the old ``config.php`` file or customize the new version.

* Run ``dbupgrade.php`` as your web server user::

    cd /your-installation-directory/include/scripts
    php dbupgrade.php

* Test your new setup.


PHP setup
---------

The following parameters are required to be set in your ``php.ini`` (these
values are defaults since PHP 5.0, but they might be different in your setup):

* ``magic_quotes_gpc``: must be "Off".
* ``magic_quotes_runtime``: must be "Off".
* ``date.timezone``: must be set to your system preference.

The maximal upload limit is determined by several PHP configuration
parameters. Check your ``php.ini`` for:

* ``file_uploads``: must be "On".
* ``upload_tmp_dir``: ensure enough space is available.
* ``upload_max_filesize``: change as needed.
* ``post_max_size``: must be at least 1M larger than upload_max_filesize.
* ``apc.rfc1867``: must be "On" for the upload progress-bar to work.
* ``session.gc_maxlifetime``: must be long enough to allow large uploads to finish.

The upload limit as shown in the submission form is determined automatically
from the ``upload_max_filesize`` parameter.

Any upload beyond ``post_max_size`` will be completely ignored: users will get
a blank page instead of an error message. You should raise ``post_max_size``
above ``upload_max_filesize`` to get an acceptable "error window".

You should also check ``session.gc_maxlifetime`` (in seconds) to be long enough
for your users to complete a large upload. Uploading 500MB on a slow ADSL
connection can take as much as 12 hours, so set it to *at least* 43200.

If you are allowed to do so, you can also set these parameters with "ini_set()"
directly inside ``include/config.php`` (so that it only affects dl). If PHP was
built as an Apache module you can also set them through ``.htaccess`` (see
http://www.php.net/manual/en/configuration.changes.php).

The upload progress-bar is only available when the PHP APC module in installed,
and the ``apc.rfc1867`` setting is explicitly enabled. The PHP APC module is
only reported to work on ``mod_php`` or ``php_fpm`` currently.


User setup
----------

DL can use both an internal and an external user database, by trusting the
authentication credentials provided by your web server.

dl supports both "normal" users and "administrators". A normal user can only
see and manage tickets created by himself. Administrators can see/manage all
the tickets.


Internal authentication
~~~~~~~~~~~~~~~~~~~~~~~

Once dl has been installed and PHP is set-up correctly, you have to create at
least one user to be able to log in. User management is handled through the
command line by using the bundled ``useradmin.php`` utility.

On the server, execute the following commands *with the same user as your web
server* (by using ``su`` or ``sudo`` if necessary)::

  cd /your-installation-directory/include/scripts
  php useradmin.php add "admin" "true" "change me"

where:

* ``admin`` is the user name
* ``true`` (or ``false``) sets the administrator status
* ``change me`` is the password

Repeat as many times as necessary. You should now be able to use the web
service. Other users/administrators can be added through the web interface.


External authentication
~~~~~~~~~~~~~~~~~~~~~~~

External authentication should be the preferred form of authentication for
corporate use since it supports whatever authentication scheme your web server
already supports (for example, LDAP, ActiveDirectory, etc).

To enable external authentication you have to protect the two files:

* ``admin.php``
* ``rest.php``

using a "Basic" authentication scheme. You should then set ''$authRealm'' to
the same authentication realm used in your web server. The other files *must
not* be protected.

DL will implicitly trust the credentials provided by the web server. All users
are logged in as "normal" by default. The only setup required is adding the
administrators with ``useradmin.php`` without using any password.

Logout with HTTP authentication is not guaranteed to work: users should simply
**close their browser** to clear their session (closing a tab or window is not
enough in many browsers). Currently, logout works as expected on:

* Firefox 3.x/4.x
* Safari 4.x
* Google Chrome/Chromium

Logout does not work on:

* Internet Explorer 7/8.
* Opera 9/10.

Again, only the *Basic* authentication is supported, which transmits the
password in clear-text unless you use SSL.


Web-server Configuration
========================

Apache/mod_php
--------------

With internal authentication::

  <Directory /your-installation-directory>
    AllowOverride Limit
    Options -Indexes
  </Directory>

With external authentication::

  <Directory /your-installation-directory>
    AllowOverride Limit
    Options -Indexes
    <FilesMatch "^(admin|rest)\.php$">
      AuthType Basic
      AuthName "Restricted Area"
      ...
      Require valid-user
    </FilesMatch>
  </Directory>

With LDAP or ActiveDirectory authentication::

  <Directory /your-installation-directory>
    AllowOverride Limit
    Options -Indexes
    <FilesMatch "^(admin|rest)\.php$">
      AuthType Basic
      AuthName "Restricted Area"
      AuthBasicProvider ldap
      AuthzLDAPAuthoritative off
      AuthLDAPURL ldap://XXXXXX:XXXX/ou=XXXX,dc=XXXX,dc=XXX?sAMAccountName?sub?(objectClass=*)
      AuthLDAPBindDN "cn=XXXX,ou=XXXXX,dc=XXX,dc=XXX"
      AuthLDAPBindPassword "XXXXX"
      ...
      Require valid-user
      Satisfy any
    </FilesMatch>
  </Directory>


Apache/FastCGI
--------------

FastCGI support in Apache up to 2.2.x is severely lacking with all the
available modules: ``mod_fcgi``, ``mod_fcgid`` (now merged officially into
Apache's ``mod_fcgi``) and ``mod_fastcgi``.

* ``mod_fcgi`` and ``mod_fcgid`` buffer the entire request in memory before
  handing-off the request to PHP, meaning that the maximal upload limit is
  bound to your available memory at the time of the request, independently of
  how PHP is setup. This is a known, old bug_ that's still present in both
  ``mod_fcgi`` 2.2.14 and ``mod_fcgid`` 2.3.4. There is no known work-around:
  either use ``mod_php`` or use a different server.

* ``mod_fastcgi`` has been proved to be slow (and sometimes unstable) in most
  configurations. It is not advisable to use PHP with ``mod_fastcgi``.

.. _bug: http://sourceforge.net/mailarchive/forum.php?thread_name=48485BDC.1020204@oxeva.fr&forum_name=mod-fcgid-users

For HTTP/External authentication to work, ``mod_rewrite`` needs to be enabled,
and a different setup is required, as shown::

  <Directory /your-installation-directory>
    AllowOverride Limit
    Options -Indexes
    <FilesMatch "^(admin|rest)\.php$">
      RewriteEngine on
      RewriteCond %{HTTP:Authorization} ^(.*)
      RewriteRule ^(.*) - [E=HTTP_AUTHORIZATION:%1]
      AuthType Basic
      AuthName "Restricted Area"
      ...
      Require valid-user
    </FilesMatch>
  </Directory>


Lighttpd
--------

PHP/FastCGI works fine with Lighttpd 1.4.x without any special setup. The
following configuration is required to protect the include directories::

  $HTTP["url"] =~ "^/dl(?:/|/.*/)include/" {
    url.access-deny = ( "" )
  }

You can also enable external authentication with the following::

  $HTTP["url"]    =~ "^/dl/(?:admin|rest)\.php$" {
    auth.require  += ( "" => (
        "method"  => "basic",
        "realm"   => "Restricted Area",
        "require" => "valid-user"
    ) )
  }


Ticket expiration
=================

Ticket expiration can be either performed internally to DL (the default), or by
using the external ``include/scripts/expire.php`` utility with a cron job. This
preference can be set by controlling the ''$gcInternal'' parameter.

The internal method requires no setup, but the external method has the added
advantage of not interrupting the web interface during the expiration process,
and also ensures that the spool is emptied when DL itself is not used actively.


Internal method
---------------

Expiration is usually performed aggressively at every page request. You can
control this behavior (thus reducing the DB pressure) by tuning the
''$gcProbability'' and ''$gcLimit'' parameters.

If you notice too much load on your DB, start by lowering ''$gcProbability'' to
0.5 and set ''$gcLimit'' to roughly the number of active tickets currently
present in your DB.

Continue to lower ''$gcProbability'' even further until the load becomes
acceptable. When the load is acceptable, but queries take too long, reduce
''$gcLimit''.


External method
---------------

Simply call ``include/scripts/expire.php`` within a cron job, which should be
executed with *the same user as the web server*. Executing the script once a
day is sufficient for low traffic websites, but can be executed as often as
needed. ''$gcLimit'' still controls how many tickets are expired for each run
to limit the execution time.


Known limitations/Browser support
=================================

* Tested with Safari 3.x/4.x, Firefox 3.x/4.x, Explorer 7/8,
  Google Chrome/Chromium.
* Tested with PHP 5.x.


Internationalization support
============================

DL is fully translated in English, German, French, Spanish and Italian. DL will
attempt to detect the correct locale of the browser and use it if available,
otherwise switch to the customizable default. The user can however switch the
language anytime.

Adding a new translation is easy for anyone familiar with the `gettext` tools:
you don't need programming knowledge. If you want to contribute a new language,
we recommend to subscribe to the mailing list and ask for guidance. We really
appreciate your help.


Contributing a new translation
------------------------------

Contributing a new translation is easy enough:

* Edit ``include/lang.php`` and add your new language name and alias to
  ``$langData``, as done for the other languages.

* Execute::

    mkdir -p include/locale/lang/LC_MESSAGES/
    cd include/scripts/
    ./langgen.php > ../locale/lang/LC_MESSAGES/messages.po

  where *lang* is the full locale name you just added.

* Translate the freshly generated ``messages.po`` using a text editor, or
  by using PoEdit_, or any other "po" editing tool.

* To test/update the translations run ``langupd.php``::

    cd include/scripts/
    ./langupd.php

* Optionally translate the user guide, which is located in
  ``include/static/guide/``.

  Copy the english directory tree into a new tree with the new locale name and
  translate ``index.txt``. ``index.html`` is regenerated automatically with
  docutils_ by running ``langupd.php`` as before.

.. _PoEdit: http://poedit.sourceforge.net/
.. _docutils: http://docutils.sourceforge.net/


Usage
=====

DL should be usable by users without any training. The web interface must be
self-explanatory. If you find the usage to be difficult or that the interface
could be improved, **it's a bug**. Please let us know.


Command-line client: ``dl-cli``
-------------------------------

A command-line client to the REST interface is included in the distribution in
``client/dl-cli.py``. This client requires a simple text configuration file, by
default stored in ``~/.dl.rc``, containing the following values:

* url: REST URL of the service
* user: your user name
* pass: your password
* verify (optional): "true" or "false": enable/disable SSL verification
  (might be required for self-signed certificates, defaults to true)

An example::

  url=https://dl.example.com/rest.php
  user=test
  pass=test

Simply run the command with no arguments to see usage information. At least
Python 2.7 is required, with the "ConfigObj" and "PycURL" modules
installed. Under Debian/Ubuntu systems you can install the required
dependencies by doing the following::

  sudo apt-get install python-pycurl python-configobj


Graphical client: ``dl-wx``
---------------------------

A graphical client is also included in the distribution, which allows to create
tickets easily from the system's taskbar. The client can be run by executing
``client/dl-wx/dl-wx.py``.

Upon first execution the user will be prompted for the basic configuration.
After that all DL functions can be operated through the taskbar icon:

* Left-clicking on the taskbar will create a new ticket using the default
  settings.
* Right-clicking allows to select different actions.
* On OSX, you can drop files directly on the dock.

At least Python 2.7 is required, with the "ConfigObj", "PycURL" and "wxPython"
modules installed. Under Debian/Ubuntu systems you can install the required
dependencies by doing the following::

  sudo apt-get install python-pycurl python-configobj python-wxgtk2.8

``dl-cli`` and ``dl-wx`` share the same configuration file, so both can be used
interchangeably.


General/support mailing list
============================

If you feel to discuss improvements, customizations and suggestions, you can
subscribe to `dl-ticket-service` by either sending an empty email to
<dl-ticket-service+subscribe@thregr.org>, using GMane_ (group
"gmane.comp.web.dl-ticket-service.general") or by contacting the author at
<wavexx@users.sf.net>. The archives are accessible via web through
http://news.gmane.org/gmane.comp.web.dl-ticket-service.general or via news
directly.

.. _GMane: http://www.gname.org/


Customisation and development
=============================

You are encouraged to change DL as you see fit under the terms of the GNU GPL
2 license. DL's GIT repository is publicly accessible at::

  git://src.thregr.org/dl


Authors and Copyright
=====================

"dl" can be found at http://www.thregr.org/~wavexx/software/dl/

"dl" is distributed under GNU GPL 2, WITHOUT ANY WARRANTY.
Copyright(c) 2007-2012 by Yuri D'Elia <wavexx@users.sf.net>.


NEWS
dl 0.10: 06/02/2012
-------------------

* The default configuration file has been renamed to "config.php.dist" and must
  now be manually copied/renamed to be used. If a suitable "config.php" is not
  found in the include/ directory, then the configuration is read from
  "/etc/dl.php". This will allow smoother release upgrades in the future.
* Notifications of tickets and grants now use the same locale that was used
  during the creation of the ticket/grant itself.
* A new "Preferences" page has been added, allowing users to change their
  password (currently supported only for internal authentication).
* Changing "hours after last download" while editing a ticket didn't actually
  extend the ticket lifetime. Changing the ticket lifetime now works correctly.
* "hours after last download" has been changed to "days after last download".
* Default ticket/grant expiration settings have been increased significantly.
* A new REST API has been implemented, allowing external applications to use the
  service programmatically.
* Two python clients have been added to the distribution: a command-line python
  client "dl-cli" and a graphical client "dl-wx".
* Minor bug and UI fixes.

Please note: DL 0.10 requires a database schema and webserver configuration
update! Please read the database upgrade procedure in the README and the
relevant notes about web server configuration.


dl 0.9.1: 31/12/2011
--------------------

* Fixed a grave security issue: unauthorized parties can perform login as any
  arbitrary user when using the built-in authentication mechanism by supplying
  an authorization header. DL versions down to 0.3 are affected.


dl 0.9: 06/04/2011
------------------

* The settings of tickets and grants are now stored independently.
* Default ticket/grant settings can be specified in the config file.
* Most ticket and grant options are now moved into an "advanced" panel.
* Grant notifications now include the ticket password in the message.
* Form validation is now also performed in JavaScript.
* Enlarged the width of the interface to 800px.
* Users management is now available through the web interface.
* French, Italian, Spanish and German translation.
* Multi-line comments can be attached to tickets and grants.
* Improved the e-mail notification text.
* Improved ticket and grant listings.
* Tickets can now be edited after being created.
* Minor bug and UI fixes.


dl 0.8: 10/07/2010
------------------

* Update PHP-Gettext to 1.1.10 (fixing several PHP Notices).
* Fix browser language autodetection (typo, thanks to Bert-Jan Kamp).
* Fix ticket expiration when using sqlite3 (table locking issues).
* Do not purge tickets prematurely after an unsuccessful download.
* Purge tickets immediately after the download, when possible.
* "useradmin.php" now allows to reset/change user role and password.
* All notifications are now sent using the default locale.
* Ticket expiration can be performed with an external utility.
* The user-guide is now included in the admin interface.


dl 0.7: 10/03/2010
------------------

* Fix XSS vulnerability for unknown ticket IDs (discovered by Sven Eric Neuz)


dl 0.6: 03/03/2010
------------------

* Remember the selected language with a cookie.
* Allow to tune the DB expiration process to improve the performance.
* Fixed E-Mail subject encoding.
* German translation update.
* PHP 5.3 warning fixes.


dl 0.5: 09/02/2010
------------------

* Fix upload progress-bar on Chrome and Safari.
* Minor bug, UI and usability fixes.
* Internationalization support.
* Italian and German translation.
* License changed to GNU GPL 2.


dl 0.4: 24/11/2009
------------------

* Ticket activity can be logged to syslog or a file.
* The minimal required PHP version is now 5.0.
* PDO is now used for the users/tickets database (defaulting to a sqlite
  database). Upgrading instructions in the README.
* The submission form now allows to automatically send a link of the ticket to
  the specified address/es.
* A ticket can now require a password to be downloaded.
* "Upload grants" can now be created, allowing others to send you a single file
  through DL.
* Progress-bar indicator during uploads.
* Required fields are highlighted when missing.


dl 0.3: 02/09/2009
------------------

* CSS-ification, with new skin from Kim Thostrup <kim@thostrup.dk>.
* Include IE5/6 PNG fix from Angus Turnbull http://www.twinhelix.com
* Multiuser support with HTTP authentication or internal user database.
* License changed to LGPL 3


dl 0.2: 10/10/2007
------------------

* Renamed "aux.php" to "funcs.php" to avoid "reserved file name" errors
  under Windows.
* Support commas in addition to semicolons as e-mail separators in the
  notify field.
* Removed the 'ID' field in "active tickets" listings.
* Allow to attach a comment in any ticket.
* Byte-ranges support.


dl 0.1: 15/06/2007
------------------

* First release.


client/dl-wx/dl-wx.py
#!/usr/bin/env python
import configobj
import validate
import argparse
import os.path
import copy
import time
import sys
import wx
import wx.xrc as xrc
from dl import *

DL_VERSION = "0.10"
DL_AGENT = "dl-wx/" + DL_VERSION
DL_DESCRIPTION = "Download Ticket Service"
DL_URL = "http://www.thregr.org/~wavexx/software/dl/"
DL_ICON = "dl-icon.ico"
DL_RC = "~/.dl.rc"

# This is rather ugly for *NIX (should use pkg_resources)
if '_MEIPASS2' in os.environ:
    RC_PATH = os.environ['_MEIPASS2']
else:
    RC_PATH = os.path.dirname(sys.argv[0])


def create_menu_item(menu, label, func, id=wx.ID_ANY):
    item = wx.MenuItem(menu, id, label)
    menu.Bind(wx.EVT_MENU, func, item)
    menu.AppendItem(item)
    return item


class TaskBarIcon(wx.TaskBarIcon):
    def __init__(self, dlapp):
        super(TaskBarIcon, self).__init__()
        self.dlapp = dlapp
        img = wx.Bitmap(os.path.join(RC_PATH, DL_ICON))
        self.SetIcon(wx.IconFromBitmap(img), DL_DESCRIPTION)
        self.Bind(wx.EVT_TASKBAR_LEFT_UP, self.dlapp.express_ticket)

    def CreatePopupMenu(self):
        menu = wx.Menu()
        create_menu_item(menu, "New Ticket", self.dlapp.new_ticket, wx.ID_NEW)
        create_menu_item(menu, "Preferences", self.dlapp.show_prefs, wx.ID_PREFERENCES)
        menu.AppendSeparator()
        create_menu_item(menu, "About", self.dlapp.show_about, wx.ID_ABOUT)
        create_menu_item(menu, "Quit", self.dlapp.quit, wx.ID_EXIT)
        return menu


class Prefs(wx.Dialog):
    def __init__(self, service, change_fn):
        self.service = service
        self.change_fn = change_fn
        self.xrc = xrc.XmlResource(os.path.join(RC_PATH, 'preferences.xrc'))
        self.PostCreate(self.xrc.LoadDialog(None, 'preferences'))
        self.url = xrc.XRCCTRL(self, 'url')
        self.username = xrc.XRCCTRL(self, 'username')
        self.password = xrc.XRCCTRL(self, 'password')
        self.verify = xrc.XRCCTRL(self, 'verify')
        self.cancel = xrc.XRCCTRL(self, 'cancel')
        self.cancel.Bind(wx.EVT_BUTTON, self.on_close)
        self.save = xrc.XRCCTRL(self, 'save')
        self.save.Bind(wx.EVT_BUTTON, self.on_save)
        self.Bind(wx.EVT_SHOW, self.on_show)
        if service.url:
            self.cancel.Show()
            self.Bind(wx.EVT_CLOSE, self.on_close)
        else:
            self.cancel.Hide()
            self.Bind(wx.EVT_CLOSE, self.on_save)

    def on_show(self, evt):
        self.set_service()

    def set_service(self):
        self.url.SetValue(self.service.url)
        self.username.SetValue(self.service.username)
        self.password.SetValue(self.service.password)
        self.verify.SetValue(self.service.verify)

    def get_service(self, service):
        service.url = self.url.GetValue().encode('utf8')
        service.username = self.username.GetValue().encode('utf8')
        service.password = self.password.GetValue().encode('utf8')
        service.verify = self.verify.GetValue()

    def on_close(self, evt=None):
        self.Hide()

    def on_save(self, evt):
        service = Service()
        self.get_service(service)
        error = None
        if not service.url:
            error = "The REST URL is mandatory"
        elif not service.username:
            error = "The username is mandatory"
        elif not service.password:
            error = "The password is mandatory"
        if error is not None:
            wx.MessageBox(error, 'Preferences', wx.OK | wx.ICON_ERROR)
        else:
            self.get_service(self.service)
            self.change_fn()
            self.on_close()


class Upload(wx.Dialog):
    def __init__(self, file, dl, params):
        self.xrc = xrc.XmlResource(os.path.join(RC_PATH, 'upload.xrc'))
        self.PostCreate(self.xrc.LoadDialog(None, 'upload'))
        self.Bind(wx.EVT_CLOSE, self.on_cancel)
        self.file = xrc.XRCCTRL(self, 'file')
        self.gauge = xrc.XRCCTRL(self, 'progress')
        self.status = xrc.XRCCTRL(self, 'status')
        self.request = dl.new_ticket(file, params, async=True,
                                     complete_fn=self.completed,
                                     failed_fn=self.failed,
                                     progress_fn=self.progress)
        self.file.SetLabel(os.path.basename(file))
        self.action = xrc.XRCCTRL(self, 'action')
        self.action.SetLabel("Cancel")
        self.action.Bind(wx.EVT_BUTTON, self.on_cancel)
        self.stamp = time.time()
        self.Fit()
        self.Show()
        self.request.start()

    def on_cancel(self, evt):
        self.status.SetLabel("Cancelling upload ...")
        self.request.cancel()

    def on_close(self, evt=None):
        self.Destroy()

    def on_progress(self, upload_t, upload_d, upload_s):
        prc = upload_d * 100 / upload_t
        ks = upload_s / 1024
        self.gauge.SetValue(prc)
        stamp = time.time()
        if stamp - self.stamp >= 1:
            self.stamp = stamp
            self.status.SetLabel("Uploading ({:.1f}%, {:.1f}KiB/s) ...".format(prc, ks))

    def progress(self, download_t, download_d, download_s, upload_t, upload_d, upload_s):
        if upload_d > 0:
            wx.CallAfter(self.on_progress, upload_t, upload_d, upload_s)

    def on_completed(self, ret):
        self.url = ret['url']
        self.status.SetLabel(self.url)
        self.action.SetLabel("Copy")
        self.action.Bind(wx.EVT_BUTTON, self.on_copy)
        self.Bind(wx.EVT_CLOSE, self.on_close)
        self.Fit()

    def completed(self, ret):
        wx.CallAfter(self.on_completed, ret)

    def on_failed(self, ex):
        if ex is None:
            self.on_close()
        else:
            error = str(ex)
            self.status.SetLabel(error)
            self.action.SetLabel("Close")
            self.action.Bind(wx.EVT_BUTTON, self.on_close)
            self.Bind(wx.EVT_CLOSE, self.on_close)
            self.Fit()
            wx.MessageBox(error, 'Upload error', wx.OK | wx.ICON_ERROR)

    def failed(self, ex):
        wx.CallAfter(self.on_failed, ex)

    def on_copy(self, evt=None):
        wx.TheClipboard.Open()
        wx.TheClipboard.SetData(wx.TextDataObject(self.url))
        wx.TheClipboard.Close()
        self.Destroy()


class NewTicket(wx.Dialog):
    def __init__(self, dl, ticket_params, change_fn):
        self.dl = dl
        self.def_ticket_params = ticket_params
        self.change_fn = change_fn
        self.xrc = xrc.XmlResource(os.path.join(RC_PATH, 'newticket.xrc'))
        self.PostCreate(self.xrc.LoadDialog(None, 'newticket'))
        self.Bind(wx.EVT_CLOSE, self.on_close)
        self.file = xrc.XRCCTRL(self, 'file')
        self.perm = xrc.XRCCTRL(self, 'perm')
        self.perm.Bind(wx.EVT_CHECKBOX, self.on_perm)
        self.total_days = xrc.XRCCTRL(self, 'total_days')
        self.days_after_dl = xrc.XRCCTRL(self, 'days_after_dl')
        self.downloads = xrc.XRCCTRL(self, 'downloads')
        self.upload = xrc.XRCCTRL(self, 'upload')
        self.upload.Bind(wx.EVT_BUTTON, self.on_upload)
        self.set_defaults = xrc.XRCCTRL(self, 'set_defaults')
        self.set_defaults.Bind(wx.EVT_BUTTON, self.on_set_defaults)
        self.ticket_params = copy.copy(self.def_ticket_params)
        self.set_ticket_params(self.ticket_params)
        self.Show()

    def on_perm(self, evt=None):
        enable = not self.perm.GetValue()
        self.total_days.Enable(enable)
        self.days_after_dl.Enable(enable)
        self.downloads.Enable(enable)

    def set_ticket_params(self, ticket_params):
        self.perm.SetValue(ticket_params.permanent)
        self.total_days.SetValue(ticket_params.total / (3600 * 24))
        self.days_after_dl.SetValue(ticket_params.lastdl / (3600 * 24))
        self.downloads.SetValue(ticket_params.downloads)
        self.on_perm()

    def get_ticket_params(self, ticket_params):
        ticket_params.permanent = self.perm.GetValue()
        ticket_params.total = self.total_days.GetValue() * 3600 * 24
        ticket_params.lastdl = self.days_after_dl.GetValue() * 3600 * 24
        ticket_params.downloads = self.downloads.GetValue()

    def on_set_defaults(self, evt):
        self.get_ticket_params(self.def_ticket_params)
        self.change_fn()

    def on_upload(self, evt):
        path = self.file.GetPath().encode('utf8')
        if not path:
            wx.MessageBox('Please select a file!', 'New Ticket', wx.OK | wx.ICON_ERROR)
        else:
            self.get_ticket_params(self.ticket_params)
            Upload(path, self.dl, self.ticket_params)
            self.on_close()

    def on_close(self, evt=None):
        self.Destroy()


class DLApp(wx.App):
    def OnInit(self):
        self.dl = DL()

        self.load_prefs()
        self.prefs = Prefs(self.dl.service, self.save_prefs)
        if not self.dl.service.url:
            wx.MessageBox('This is the first time you run ' + DL_DESCRIPTION +
                          '. You need to configure it first.',
                          'Preferences', wx.OK | wx.ICON_INFORMATION)
            self.prefs.ShowModal()

        self.tbi = TaskBarIcon(self)
        stub = wx.Frame(None)
        menu = wx.MenuBar()
        menu.Append(self.tbi.CreatePopupMenu(), "&File")
        stub.SetMenuBar(menu)
        self.SetTopWindow(stub)
        return True

    def load_prefs(self):
        cfgpath = os.path.expanduser(DL_RC)
        self.cfg = configobj.ConfigObj(cfgpath)
        v = validate.Validator()

        self.dl.service.url = v.check('string', self.cfg.get('url', ''))
        self.dl.service.username = v.check('string', self.cfg.get('user', ''))
        self.dl.service.password = v.check('string', self.cfg.get('pass', ''))
        self.dl.service.verify = v.check('boolean', self.cfg.get('verify', True))
        self.dl.service.agent = DL_AGENT

        self.ticket_params = TicketParams()
        self.ticket_params.permanent = v.check('boolean', self.cfg.get('perm', False))
        self.ticket_params.total = v.check('integer', self.cfg.get('total_days', 365)) * 3600 * 24
        self.ticket_params.lastdl = v.check('integer', self.cfg.get('days_after_dl', 30)) * 3600 * 24
        self.ticket_params.downloads = v.check('integer', self.cfg.get('downloads', 0))

    def save_prefs(self):
        self.cfg['url'] = self.dl.service.url
        self.cfg['user'] = self.dl.service.username
        self.cfg['pass'] = self.dl.service.password
        self.cfg['verify'] = self.dl.service.verify
        self.cfg['perm'] = self.ticket_params.permanent
        self.cfg['total_days'] = self.ticket_params.total / (3600 * 24)
        self.cfg['days_after_dl'] = self.ticket_params.lastdl / (3600 * 24)
        self.cfg['downloads'] = self.ticket_params.downloads
        self.cfg.write()

    def express_ticket(self, evt=None):
        path = wx.FileSelector(flags=wx.FD_OPEN|wx.FD_FILE_MUST_EXIST).encode('utf8')
        if path:
            Upload(path, self.dl, self.ticket_params)

    def new_ticket(self, evt=None):
        NewTicket(self.dl, self.ticket_params, self.save_prefs)

    def show_prefs(self, evt=None):
        self.prefs.Show()

    def show_about(self, evt=None):
        wx.MessageBox(DL_DESCRIPTION + " " + DL_AGENT + "\n" + DL_URL,
                      'About', wx.OK | wx.ICON_INFORMATION)

    def quit(self, evt=None):
        self.ExitMainLoop()

    def MacOpenFile(self, path):
        Upload(path.encode('utf8'), self.dl, self.ticket_params)

    def MacReopenApp(self):
        self.express_ticket()


if __name__ == '__main__':
    main = DLApp()
    main.MainLoop()
